presto-cli/src/main/java/io/prestosql/cli/OutputHandler.java

```diff
@@ -13,85 +13,95 @@
  */
 package io.prestosql.cli;
 
-import io.airlift.units.Duration;
+import com.google.common.collect.Queues;
 import io.prestosql.client.StatementClient;
 
 import java.io.Closeable;
 import java.io.IOException;
 import java.util.ArrayList;
 import java.util.List;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.LinkedBlockingQueue;
 import java.util.concurrent.atomic.AtomicBoolean;
 
-import static io.airlift.units.Duration.nanosSince;
 import static java.util.Collections.unmodifiableList;
 import static java.util.Objects.requireNonNull;
 import static java.util.concurrent.TimeUnit.SECONDS;
 
 public final class OutputHandler
         implements Closeable
 {
-    private static final Duration MAX_BUFFER_TIME = new Duration(3, SECONDS);
-    private static final int MAX_BUFFERED_ROWS = 10_000;
+    private static final int MAX_QUEUED_ROWS = 50_000;
+    private static final int MIN_BUFFERED_ROWS = 100;
+    private static final long MAX_BUFFER_TIMEOUT = 3;
 
     private final AtomicBoolean closed = new AtomicBoolean();
-    private final List<List<?>> rowBuffer = new ArrayList<>(MAX_BUFFERED_ROWS);
+    private final BlockingQueue<List<?>> rowQueue = new LinkedBlockingQueue<>(MAX_QUEUED_ROWS);
```

Comment:
The section `Test case #4 – Adding and removing elements` in [Queue battle and the Linked ConcurrentHashMap](https://www.javacodegeeks.com/2010/09/java-best-practices-queue-battle-and.html) has the performance comparison of ArrayBlockingQueue and LinkedBlockingQueue, it says that LinkedBlockingQueue is the best choice for producer-consumer scenarios.
Also, [when to prefer LinkedBlockingQueue over ArrayBlockingQueue](https://stackoverflow.com/questions/35967792/when-to-prefer-linkedblockingqueue-over-arrayblockingqueue) mentions LinkedBlockingQueue should have higher throughput because of separate locks.
I think our use of BlockingQueue will fall in the same bucket. Please let me know if we need to change it to ArrayBlockingQueue.

Reply:
Interesting, thanks for the references. For the *Queue battle*, they are using 50 worker threads, whereas we have a single producer and consumer. My suspicion is that constantly allocating nodes for the linked list would have overhead, but maybe it doesn't matter.

There's an easy solution: just try out both and see if there is any difference in performance. It's possible the choice of queue is irrelevant to the overall performance, or maybe it has a big impact -- need to try it.

Reply:
The ArrayBlockingQueue turned out to be more performant in the experiment.
Thanks for the suggestion, I have made the changes.

Reply:
Did we really change it back to ArrayBlockingQueue? I still see LinkedBlockingQueue in the codebase. 

https://github.com/prestosql/presto/blob/2e99be478672ff04f52cc003cf64c947ef140b15/presto-cli/src/main/java/io/prestosql/cli/OutputHandler.java#L42

Reply:
@sajjoseph you're right. Looks like unintentional omission. cc @ShashwatArghode

Reply:
Thanks @sajjoseph and @findepi for spotting it, I made ArrayBlockingQueue changes in PrestoResultSet and missed this one in OutoutHandler. I will make the changes and cross-check that I haven't missed anything else.